New framing section for Chapter 16

pavpanchekha opened the pull request:
This PR adds a new "framing" section for Chapter 16. I also wrote a "go further" based on some comments in our outline that I couldn't fit in anywhere, but it would be nice if there were some useful links we could put in there.

chrishtr commented on book/invalidation.md:
```diff
@@ -167,6 +167,79 @@ API!
+=================
+
+Fundamentally, the reason editing this page is slow in our browser is
+that this page is pretty big.[^other-reasons] After all, it's not
```
```suggestion
it is pretty big.[^other-reasons] After all, it's not
```

pavpanchekha replied:
I did something different but I think still avoids the duplication.

chrishtr commented on book/invalidation.md:
```diff
@@ -167,6 +167,79 @@ API!
+that this page is pretty big.[^other-reasons] After all, it's not
+handling the keypress that's slow---that just appends a character to a
+`Text` node, which takes almost no time. It's what happens afterwards
+that takes time: the browser has to rerender the whole page. Even tiny
```
```suggestion
that takes time: the browser has to rerender the whole page. Even seemingly tiny
```

chrishtr commented on book/invalidation.md:
```diff
@@ -167,6 +167,79 @@ API!
+    that simple optimizations won't achieve.
+
+If we want interactions to be fast, even on large, complex pages,
+we're going to need re-rendering the page to take time *proportional
```
```suggestion
we're going to need re-rendering the page to take time proportional
```

chrishtr commented on book/invalidation.md:
```diff
@@ -167,6 +167,79 @@ API!
+
+If we want interactions to be fast, even on large, complex pages,
+we're going to need re-rendering the page to take time *proportional
+to the size of the change*, not proportional to the size of the page.
```
```suggestion
to the *size of the change*, and *not* proportional to the size of the page.
```

chrishtr commented on book/invalidation.md:
```diff
@@ -167,6 +167,79 @@ API!
+If we want interactions to be fast, even on large, complex pages,
+we're going to need re-rendering the page to take time *proportional
+to the size of the change*, not proportional to the size of the page.
+This property is sometimes called *incrementality*, and it really
```
I don't think I have heard this name, maybe it's a thing in other communities?
Anyway, I think we can coin it and call it the *principle of incremental performance.*

Here would be a good place to explain why this is important:
for developers, so that they can predict the performance
of the web pages they create, and important for users, so they can predict the
performance of interactions (typing: fast; loading a new page: relatively slow).
We can then follow up with commentary about how
the inversion-of-control/constraint programming model of HTML and CSS makes it
hard to achieve this principle, and this chapter is all about how to do so.

pavpanchekha replied:
Great name!

chrishtr commented on book/invalidation.md:
```diff
@@ -167,6 +167,79 @@ API!
+going to take time proportional to the size of the page. So to achieve
+incrementality, we're going to need to think of the initial render and
+later re-renders differently. The initial render will create a cache
+that later renders invalidate and recompute parts of.[^big-change] If
```
```suggestion
that later renders invalidates and recompute parts of.[^big-change] If
```

chrishtr commented on book/invalidation.md:
```diff
@@ -167,6 +167,79 @@ API!
+later re-renders differently. The initial render will create a cache
+that later renders invalidate and recompute parts of.[^big-change] If
+we're only invalidate a small part each time, this will only take time
+proportional to the changed portion of the page.
```
This sentence feels like it's repeating earlier ones.

chrishtr commented on book/invalidation.md:
```diff
@@ -167,6 +167,79 @@ API!
+layout tree and traversing it to compute layout fields. Ultimately,
+re-layouts will take under a millisecond when typing on this page.
+
+Getting there is going to be a bit of a rocky road, though. The whole
```
Instead of stating generalities about the hard work to come, being more explicit about it?

Something like:

"We'll have to proceed with the following tasks:
* Make a detailed *dependency graph* of what depends on what, down to the level of every
value (`width`, `height`, `y`, etc) on every layout object.
* Use that graph to determine what to recompute when something changes.

chrishtr commented on book/invalidation.md:
```diff
@@ -167,6 +167,79 @@ API!
+and for you, to see how real browsers make interactive pages fast.
+
+::: {.further}
+Implementing incremental layout is complex, but it's a core part of
```
This could be rephrased to refer back to the "principle of incremental performance", and focus
on why it's important for browsers to provide this as a core value proposition to developers and users.
(Above I'm suggesting that we focus on the *fact* that we'll be focusing on that, but here we can talk about
why -- having to do with the role of a platform provider in a layered platform: platform engineer vs web developer vs user.

pavpanchekha commented:
I think I addressed all of the comments, which prompted a big reorganization of the second half of the section. It's probably best to review by rereading it.

chrishtr commented on book/invalidation.md:
```diff
@@ -167,6 +167,86 @@ API!
+that it's pretty big.[^other-reasons] After all, it's not handling the
+keypress that's slow---that just appends a character to a `Text` node,
+which takes almost no time. It's what happens afterwards that takes
+time: the browser has to rerender the whole page. Even seemlingly tiny
```
```suggestion
time: the browser has to rerender the whole page. Even seemingly tiny
```

chrishtr commented on book/invalidation.md:
```diff
@@ -167,6 +167,86 @@ API!
+
+But the principle of incremental performance also really constrains
+our browser. For example, during rendering even *traversing* the whole
+page would take time proportional to the whole page, not the change
```
```suggestion
layout tree would take time proportional to the whole page, not the change
```

chrishtr commented on book/invalidation.md:
```diff
@@ -167,6 +167,86 @@ API!
+building the layout tree and traversing it to compute layout fields.
+Re-layout for this page will take under a millisecond.
+
+[^why-layout]: Layout is important and also complex enough to
```
```suggestion
[^why-layout]: Why, you might ask? Because layout is complex enough to
```

chrishtr commented on book/invalidation.md:
```diff
@@ -167,6 +167,86 @@ API!
+
+In a real browser, every step of the rendering pipeline needs to be
+incremental, but this chapter focuses on layout.[^why-layout] Most of
+the chapter is about tracking dependencies, and building abstractions
```
```suggestion
this chapter is about tracking dependencies in the dependency graph, and building abstractions
```

chrishtr commented on book/invalidation.md:
```diff
@@ -167,6 +167,86 @@ API!
+large refactor of our layout engine. But ultimately, incrementalizing
+layout will allow us to skip the two most expensive parts of layout:
+building the layout tree and traversing it to compute layout fields.
+Re-layout for this page will take under a millisecond.
```
```suggestion
When we're done re-layout for this page will take under a millisecond when making small changes,
no matter how large the page!
```

chrishtr commented on book/invalidation.md:
```diff
@@ -167,6 +167,86 @@ API!
+browser engineers, that creates a whole bunch of complexity. But think
+about the web as a whole---it involves not just browser engineers, but
+web developers and users as well. Implementing complex invalidation
+algorithms in the browser lets web developers make more interesting
```
```suggestion
algorithms in the browser lets web developers focus on making more interesting
```
